torch/csrc/jit/passes/symbolic_shape_analysis.cpp

```diff
@@ -86,54 +103,105 @@ struct SymbolicShapeAnalyzer {
+      auto tensor_value = node_->inputs().at(tensor_index);
+      auto tensor_shape =
+          tensor_value->type()->expect<TensorType>()->symbolic_sizes();
+      TORCH_INTERNAL_ASSERT(tensor_shape.rank().has_value());
+
+      for (const auto& use : graph_->inputs().at(tensor_index)->uses()) {
+        // TODO: either decompose composite ops like slice or add handling here
+        switch (use.user->kind()) {
+          case aten::len: {
+            size_t len = tensor_shape.rank().value();
+            replaceWithIValue(use.user->output(), static_cast<int64_t>(len));
+          } break;
+          case aten::__getitem__: {
+            auto index = constant_as<int64_t>(use.user->inputs().at(1));
+            if (index) {
+              auto norm_index = normIndex(*index, *tensor_shape.rank());
+              if (norm_index &&
+                  (tensor_shape[*norm_index].is_static() ||
+                   substitute_symbolic_dims)) {
+                replaceWithIValue(
+                    use.user->output(), tensor_shape[*norm_index].value());
+              } else if (norm_index) {
+                int64_t symbolic_index = tensor_shape[*norm_index].value();
+                symbolic_shape_map[symbolic_index].push_back(
+                    use.user->output());
+              }
+            }
+          }
+        }
+      }
+
+      for (const auto& symbolic_set : symbolic_shape_map) {
+        mergeSymbolicShapeSets(symbolic_set.second);
+      }
     }
   }
 
-  void substituteTensorProperties(int64_t node_input_tensor_index) {
-    auto node_tensor_index_value = node_->input(node_input_tensor_index);
-    auto shape =
-        node_tensor_index_value->type()->expect<TensorType>()->symbolic_sizes();
-    if (!shape.rank().has_value()) {
-      return;
+  bool isDominatedBy(Node* node, Node* dominator) {
+    while (node) {
+      if (node->owningBlock() == dominator->owningBlock()) {
+        return dominator->isBefore(node);
+      }
+      node = node->owningBlock()->owningNode();
     }
+    return false;
+  }
 
-    for (const auto& use :
-         graph_->inputs().at(node_input_tensor_index)->uses()) {
-      // TODO: either decompose composite ops like slice or add handling here
-      switch (use.user->kind()) {
-        case aten::len: {
-          size_t len = shape.rank().value();
-          replaceWithIValue(use.user->output(), static_cast<int64_t>(len));
-        } break;
-        case aten::__getitem__: {
-          auto index = constant_as<int64_t>(use.user->inputs().at(1));
-          if (index) {
-            auto norm_index = normIndex(*index, *shape.rank());
-            // TODO: HANDLE non-static value (symbolic shape)
-            if (norm_index && shape[*norm_index].is_static()) {
-              replaceWithIValue(
-                  use.user->output(), shape[*norm_index].static_size());
-            }
-          }
+  void mergeSymbolicShapeSets(const std::vector<Value*>& symbolic_set) {
```

Comment:
Do not understand what this tries to do, could you add more comments to explain?